voorhoede/head-start
# feature: DatoCMS collection helper

# Changes

- Adds a `datocmsCollection` helper function to fetch all records in a DatoCMS collection.
- Uses pagination so even if a collection has more than 100 records (DatoCMS query limit) it will still return all records.
- Is flexible enough for different use cases, by accepting a GraphQL fragment as an argument to use for the given collection.
- Extends `datocmsRequest` with retry on rate limit error.

Simplified example (without types):

```ts
import { datocmsCollection } from '@lib/datocms';

const pages = await datocmsCollection({ 
  collection: 'Pages',
  fragment: `slugs: _allSlugLocales { locale, value }`,
});
```

# Associated issue

Resolves #74 

# How to test

1. Checkout this branch locally
2. Force pagination, by changing `recordsPerPage` to `1` in the new `datocmsCollection` function in `lib/datocms.ts`
3. Run a local build
4. Verify all the pages are still generated in `dist/[locale]/`
5. Verify all the pages are also still generated in the Cloudflare deploy preview

# Checklist

- ~~I have added/updated tests to prove that my feature works (if not possible please explain why)~~
- ~~I have made changes to the README and if the change affects the project setup (npm commands changed, new service added, environmental variable added)~~ will add this to `docs/data-loading.md` (see #71)
